### ISSUE TITLE
arbitrary self types causes unexpected error messages

### ISSUE THREAD

**Issue by nikomatsakis** [rust-lang/rust#69069]
This example:

```rust
#![feature(arbitrary_self_types)]
trait Test<T: core::ops::Deref<Target = Self>> {
    fn is_some(self: T);
}

fn f() {
    let x = Some(2);
    if x.is_some() {
        println!("Some");
    }
}
```
[Playground link](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=83f8fd48766d3bcd61212a70b6db86df)

gives the error 

```
error[E0277]: the trait bound `std::option::Option<{integer}>: std::ops::Deref` is not satisfied
 --> src/lib.rs:8:10
  |
8 |     if x.is_some() {
  |          ^^^^^^^ the trait `std::ops::Deref` is not implemented for `std::option::Option<{integer}>`

error[E0308]: mismatched types
 --> src/lib.rs:8:8
  |
8 |     if x.is_some() {
  |        ^^^^^^^^^^^ expected `bool`, found `()`

error: aborting due to 2 previous errors

```

rather than an error because the `Test` trait is not implemented.

This is presumably just a case where improved diagnostics would be good. The [inherent `is_some` method](https://doc.rust-lang.org/std/option/enum.Option.html#method.is_some) winds up not triggering because it requires auto-ref rather than being passed "by value".

_Originally posted by @tguser402 in https://github.com/rust-lang/rust/issues/66312#issuecomment-568805751_

**Comment by estebank** [comment]
This shouldn't be marked as `[requires-nightly]` because the same problem happens with `Pin`, which _is_ stable.

**Comment by estebank** [comment]
Cc https://github.com/rust-lang/rust/issues/66289

**Comment by estebank** [comment]
We currently provide a suggestion that is missing parens to be fully correct:

```
error[[E0277]](https://doc.rust-lang.org/nightly/error_codes/E0277.html): the trait bound `Option<{integer}>: Deref` is not satisfied
 --> src/lib.rs:8:10
  |
8 |     if x.is_some() {
  |          ^^^^^^^ the trait `Deref` is not implemented for `Option<{integer}>`
  |
note: required by a bound in `Test::is_some`
 --> src/lib.rs:2:15
  |
2 | trait Test<T: core::ops::Deref<Target = Self>> {
  |               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `Test::is_some`
3 |     fn is_some(self: T);
  |        ------- required by a bound in this associated function
help: consider borrowing here
  |
8 |     if &x.is_some() {
  |        +
8 |     if &mut x.is_some() {
  |        ++++

error[[E0308]](https://doc.rust-lang.org/nightly/error_codes/E0308.html): mismatched types
 --> src/lib.rs:8:8
  |
8 |     if x.is_some() {
  |        ^^^^^^^^^^^ expected `bool`, found `()`
```
